core/shared/src/main/scala/fs2/internal/FreeC.scala

```diff
@@ -12,138 +12,204 @@ import scala.util.control.NonFatal
-    override def toString: String = s"FreeC.Pure($r)"
+
+  def pure[F[_], A](a: A): FreeC[F, A] = Result.Pure(a)
+
+  def eval[F[_], A](f: F[A]): FreeC[F, A] = Eval(f)
+
+  def raiseError[F[_], A](rsn: Throwable): FreeC[F, A] = Result.Fail(rsn)
+
+  def interrupted[F[_], A](scopeId: Token, failure: Option[Throwable]): FreeC[F, A] =
+    Result.Interrupted(scopeId, failure)
+
+  sealed trait Result[+R] { self =>
+
+    def asFreeC[F[_]]: FreeC[F, R] = self.asInstanceOf[FreeC[F, R]]
+
+    def asExitCase: ExitCase[Throwable] = self match {
+      case Result.Pure(_)           => ExitCase.Completed
+      case Result.Fail(err)         => ExitCase.Error(err)
+      case Result.Interrupted(_, _) => ExitCase.Canceled
+    }
+
+    def covary[R2 >: R]: Result[R2] = self.asInstanceOf[Result[R2]]
+
+    def recoverWith[R2 >: R](f: Throwable => Result[R2]): Result[R2] = self match {
+      case Result.Fail(err) =>
+        try { f(err) } catch { case NonFatal(err2) => Result.Fail(CompositeFailure(err, err2)) }
+      case _ => covary[R2]
+    }
+
+  }
+
+  object Result {
+
+    val unit: Result[Unit] = pure(())
+
+    def pure[A](a: A): Result[A] = Result.Pure(a)
+
+    def raiseError[A](rsn: Throwable): Result[A] = Result.Fail(rsn)
+
+    def interrupted[A](scopeId: Token, failure: Option[Throwable]): Result[A] =
+      Result.Interrupted(scopeId, failure)
+
+    def fromEither[F[_], R](either: Either[Throwable, R]): Result[R] =
+      either.fold(Result.Fail(_), Result.Pure(_))
+
+    def unapply[F[_], R](freeC: FreeC[F, R]): Option[Result[R]] = freeC match {
+      case r @ Result.Pure(_)           => Some(r: Result[R])
+      case r @ Result.Fail(_)           => Some(r: Result[R])
+      case r @ Result.Interrupted(_, _) => Some(r: Result[R])
+      case _                            => None
+    }
+
+    final case class Pure[F[_], R](r: R) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Pure($r)"
+    }
+
+    final case class Fail[F[_], R](error: Throwable) extends FreeC[F, R] with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Fail($error)"
+    }
+
+    // finalizerErrors contains accumulated errors during finalization of interruption.
+    // they will be presented to freeC after interupted scope will be closed
+    final case class Interrupted[F[_], R](scope: Token, finalizerErrors: Option[Throwable])
+        extends FreeC[F, R]
+        with Result[R] {
+      override def translate[G[_]](f: F ~> G): FreeC[G, R] =
+        this.asInstanceOf[FreeC[G, R]]
+      override def toString: String = s"FreeC.Interrupted($scope, ${})"
+    }
+
+    implicit class ResultSyntax[R](val self: Result[R]) extends AnyVal {
+
+      def map[R2](f: R => R2): Result[R2] = self match {
```

Comment:
I don't think this will ever be called b/c `map` is defined directly on `FreeC`. Hence, scalac will never look for this extension method.

Reply:
actually the problem is that `map` in FreeC is now implemented with this `map` and they both have different signatures (return types). This is picked over one on FreeC when FreeC is result. We could perhaps move `map` implementation to each FreeC case instead this magic?

Reply:
Ah, so `Result[A]` is *not* a subtype of `FreeC[F, A]`, which is why scalac doesn't see the `map` method on `FreeC`. This feels pretty delicate to me -- I think it would be clearer to override map on each instance but will defer to your opinion here.

Reply:
ok

Reply:
I prefer the current definition of `map` -- I don't like adding the extra type param (`F`) and having result type be more general than `Result[B]`.

Reply:
the trouble is that for example Pure : 

Pure[A] extends FreeC[F, A] which requires to have map[B](f: A => B): FreeC[F, B]

but also 

Pure[A] extends Result[A] which requires to have map[B](f: A=>B) : Result[B] 

and compiler doesn't obviously like that. 

Not sure if I know the solution or understand what do you mean ....

Reply:
Ah yes, okay I understand. We could fix by making `FreeC` covariant in both type params and making `Result[A] <: FreeC[Nothing, A]` but that's a lot of annoying machinery.

How about we rename `map` on `Result` to something else? `mapR` or something?

Reply:
yeah, we could. Or you know, I could simply add monad instance for result and call F.map and thats it ?